llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/IREEPyDM/IR/IREEPyDMOps.td

```diff
@@ -0,0 +1,521 @@
+    form and are progressively refined to something more specific as more
+    information is known. In general, for well-formed programs, there should
+    not be any `dynamic_call`s left at the lowest levels of the compiler.
+  }];
+
+  let arguments = (ins FlatSymbolRefAttr:$callee,
+                       Variadic<IREEPyDM_AnyValueType>:$operands);
+  let results = (outs
+      IREEPyDM_ExceptionResultType:$exc_result,
+      IREEPyDM_ObjectType:$result);
+
+  let assemblyFormat = [{
+    $callee `(` $operands `)` attr-dict `:` functional-type($operands, results)
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Boxing/unboxing
+//===----------------------------------------------------------------------===//
+
+def IREEPyDM_BoxOp : IREEPyDM_PureOp<"box"> {
+  let summary = "Boxes a primitive into an object";
+  let description = [{
+    Given a PrimitiveType, boxes it into an appropriate !object, establishing
+    identity. For value typed primitives, it is always safe to box, but for
+    reference primitives, the providence must be tracked and the original boxed
+    value used (vs boxing a new one). Failure to do so will result in aliased
+    objects.
+  }];
+
+  let arguments = (ins
+    IREEPyDM_AnyPrimitiveType:$primitive
+  );
+  let results = (outs
+    IREEPyDM_ObjectType:$object
+  );
+
+  let assemblyFormat = [{
+    $primitive `:` type($primitive)  `->` type($object) attr-dict
+  }];
+}
+
+def IREEPyDM_UnboxOp : IREEPyDM_PureOp<"unbox"> {
+  let summary = "Unboxes an object to a specific primitive type";
+  let description = [{
+    Unboxes to a primitive, returning a failure result and a default initialized
+    primitive value on failure to unbox.
+  }];
+
+  let arguments = (ins
+    IREEPyDM_ObjectType:$object
+  );
+  let results = (outs
+    IREEPyDM_ExceptionResultType:$status,
+    IREEPyDM_AnyPrimitiveType:$primitive
+  );
+
+  let assemblyFormat = [{
+    $object `:` type($object)  `->` type($primitive) attr-dict
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Name access
+// While the CPython nomenclature is a bit inconsistent, we follow it where
+// it makes sense. Names are split into a few categories:
+//   - Free Variables: Local variables that are not visible outside of the
+//     containing (actual, not lexical) function. CPython bytecode refers
+//     to operations on these as "FAST" (i.e. LOAD_FAST, STORE_FAST, DEL_FAST).
+//   - Cell Variables: Variables that exist as part of the function closure
+//     and are resolved through a level of indirection (cell).
+//   - Global Variables: Variables that are resolved via the function's
+//     globals().
+// At this level, all operations are done by name (at the CPython instruction
+// level, it is by ordinal).
+//===----------------------------------------------------------------------===//
+
+def IREEPyDM_LoadFreeVarOp : IREEPyDM_PureOp<"load_free_var"> {
```

Comment:
idea: maybe you you could express this via IR more strongly? for example:
```mlir
pydm.func @foo() {
  %free_var_a = pydm.free_var "a"
  %0 = pydm.load_free_var %free_var_a : ...
  pydm.store_free_var %0, %free_var_a : ...
```
This way if you want to perform usage analysis, hoisting, etc you have something to root onto, you can move the variables between scopes as you have them, etc. Verification also works normally (no need for lookups), inlining functions works (today you couldn't safely inline pydm.funcs as the free_vars attrs would need to be fixed, etc). You can also bake in the logic you have about NotInitialized by having a `pydm.uninitialized_free_var` vs just relying on the string attr here not being present in the function attrs list.

Another idea preserving the behavior you have would be to make the func a symbol table, such that you could do:
```mlir
pydm.func @foo() {
  pydm.free_var @a
  %0 = pydm.load_free_var @a : ...
  pydm.store_free_var %0, @a : ...
```
Then you still get DCE and can rely on built in symbol usage verification but you wouldn't get trivial inlining as with the above (would need to unique/rename the symbols, something the inliner doesn't do today).

The same approaches may work with the cell variables too, and being able to do things like promote cell vars to free vars would become natural (and safe) at this level, simplifying the subsequent python-semantic type refinement you want to do. You can also extend the SSA approach with memory effects to get some usability improvements there in your passes and - whenever someone implements them - optimizations that'll make your other transformations more robust (like load-store forwarding, redundant store elision, etc that'll help with type refinement). It also gives you a place to store a refined type for the free var when you get there and verify/coerce at all load/store sites without needing magic/expensive analysis.

I think the SSA approach would be as easy to lower into as you could just emit duplicate `pydm.free_var` ops on each use and let CSE dedupe them and carry them up to the entry block.

Reply:
Nice idea: I added a TODO to LoadFreeVarOp. It was on my list to fiddle with how to do inlining of PyFuncs (which will involve some level of symbol isolation). Thanks for the suggestion. Will figure it out in a follow-up.